site/content/en/docs/Integration Patterns/high-density-gameservers.md

```diff
@@ -0,0 +1,79 @@
+
+Utilising the new allocation `gameServerState` filter as well as the existing ability to edit the 
+`GameServer` Labels at both [allocation time]({{% ref "/docs/Reference/gameserverallocation.md" %}}), and from 
+within the game server process, [via the SDK]({{% ref "/docs/Guides/Client SDKs/_index.md#setlabelkey-value" %}}), 
+means Agones is able to atomically remove a `GameServer` from the list of potentially allocatable 
+`GameServers` at allocation time, and then return it back into the pool of allocatable `GameServers` if and when the 
+game server process deems that is has room to host another game session. 
+
+<a href="../../../diagrams/high-density.puml.png" target="_blank">
+<img src="../../../diagrams/high-density.puml.png" alt="High Density Allocation Diagram" />
+</a>
+
+{{< alert title="Info" color="info">}}
+To watch for Allocation events, there is the initial `GameServer.status.state` change from `Ready` to `Allocated`,
+but it is also useful to know that the value of `GameServer.metadata.annotations["agones.dev/last-allocated"]` will
+change with each allocation, regardless of if there is a state change or not.
+{{< /alert >}}
+
+## Example `GameServerAllocation`
+
+The below `Allocation` will first attempt to find a `GameServer` from the `Fleet` `simple-udp` that is already 
+Allocated and also has the label `agones.dev/sdk-available` with the value of `true`.
+
+The above condition indicates that the matching game server process behind the matched `GameServer` record is able to 
+accept another game session at this time.
+
+If the above Allocated `GameServer` does not exist, then allocated a Ready `GameServer` from the `simple-udp` `Fleet`.
+
+Whichever condition is met, once allocation is made against a `GameServer`, its label of `agones.dev/sdk-available` 
+will be set to the value of `false`, thereby removing it from any future allocations with the below schema.
+
+It will then be up to the game server process to decide on if and when it is appropriate to set the 
+`agones.dev/sdk-available` value back to `true`, thereby indicating that it can accept another concurrent 
+gameplay session. 
+
+```yaml
+apiVersion: "allocation.agones.dev/v1"
+kind: GameServerAllocation
+spec:
+  selectors:
+    - matchLabels:
+        agones.dev/fleet: simple-udp
+        agones.dev/sdk-available: "true" # this is important
+      gameServerState: Allocated # new state filter: allocate from Allocated servers
+    - matchLabels:
+        agones.dev/fleet: simple-udp
+      gameServerState: Ready # Allocate out of the Ready Pool (which would be default, so backward compatible)
+  metadata:
+    labels:
+      agones.dev/sdk-available: "false" # this removes it from the pool
+```
+
+{{< alert title="Info" color="info">}}
+It's important to note that the labels that the `GameServer` process use to add itself back into the pool of 
+allocatable instances, must start with the prefix `agones.dev/sdk-`, since only labels that have this prefix are 
```

Comment:
This prefix is a bit unfortunate in this example. When I first read the label, it sounds like "the sdk is available" instead of "a label set via the sdk indicates that the game server is saying it is available". 

Not that it's worth changing, but I'm curious why you did a trailing `-` instead of a trailing `/` here. Both would provide a namespace for the labels.

Reply:
> Not that it's worth changing, but I'm curious why you did a trailing - instead of a trailing / here. Both would provide a namespace for the labels.

In retrospect it would have been better with a `/`, but it was just a decision that got made early that stuck.

> When I first read the label, it sounds like "the sdk is available" instead of "a label set via the sdk indicates that the game server is saying it is available".

Since the label can be anything (best name I could come up with was `available`), if there is a better name you can think of, definitely open to suggestions!

Reply:
It's verbose, but I wonder if something like `agones.dev/sdk-gs-ready-for-allocation` or `agones.dev/sdk-gs-ready-for-reallocation` or `agones.dev/sdk-gs-ready-for-new-session` might be clearer.

Reply:
I like where this is going. I'm a fan of `agones.dev/sdk-gs-ready-for-new-session` - mainly because it's clear that it's a game session that is available within the current game server.

WDYT of `agones.dev/sdk-gs-session-available` ? Not _quite_ as verbose (35 chars vs 39) , but I think hits all the marks?

Reply:
I also like a variation on this type of label, because then SDK command becomes: `SDK.SetLabel("gs-session-available", "true")` - which I think makes more semantic sense 👍🏻

Reply:
I was leaning towards using "ready" in the label since it matches the terminology we use already (ready game servers) to indicate that you were moving into the state where you expected an allocation to come along. But I like your suggestion just as well. In some ways it's better to have a different word since this isn't an agones state and this page talks about "game sessions" instead of "game servers".

Reply:
> I was leaning towards using "ready" in the label

Oooh, I see where you are going there.  I like the bringing together or known concepts 👍🏻 

`SDK.SetLabel("gs-session-ready", "true")` (rather than available) ?

Which would become `agones.dev/sdk-gs-session-ready = true` WDYT?

Reply:
Sounds great.

Reply:
Re-reading the new version with this label, I like it even more than I thought I would. :)